src/query/server/server.go

```diff
@@ -230,77 +299,57 @@ func Run(runOpts RunOptions) {
 	if n := namespaces.NumAggregatedClusterNamespaces(); n > 0 {
 		logger.Info("configuring downsampler to use with aggregated cluster namespaces",
 			zap.Int("numAggregatedClusterNamespaces", n))
-		downsampler = newDownsampler(logger, clusterManagementClient,
+		downsampler, err = newDownsampler(clusterManagementClient,
 			fanoutStorage, instrumentOptions)
+		if err != nil {
+			return nil, nil, nil, nil, err
+		}
 	}
 
-	engine := executor.NewEngine(fanoutStorage)
-
-	handler, err := httpd.NewHandler(fanoutStorage, downsampler, engine,
-		clusterClient, cfg, runOpts.DBConfig, scope)
-	if err != nil {
-		logger.Fatal("unable to set up handlers", zap.Any("error", err))
-	}
-	handler.RegisterRoutes()
-
-	listenAddress, err := cfg.ListenAddress.Resolve()
-	if err != nil {
-		logger.Fatal("unable to get listen address", zap.Error(err))
-	}
-
-	logger.Info("starting server", zap.String("address", listenAddress))
-	go func() {
-		if err := http.ListenAndServe(listenAddress, handler.Router); err != nil {
-			logger.Fatal("unable to serve on listen address",
-				zap.String("address", listenAddress), zap.Error(err))
+	cleanup := func() error {
+		lastErr := storageCleanup()
+		// Don't want to quit on the first error since the full cleanup is important
+		if err := clusters.Close(); err != nil {
+			if lastErr == nil {
+				lastErr = err
+			} else {
+				// Make sure the previous error is at least logged
+				logger.Error("error during cleanup", zap.Error(lastErr))
```

Comment:
You might want to do this regardless? Otherwise won't log the error if storageCleanup fails but clusters close correctly